src/desktop/space/mod.rs

```diff
@@ -666,3 +670,54 @@ where
 
     damage_tracked_renderer.render_output(renderer, age, &*render_elements, clear_color, log)
 }
+
+/// Find the primary output for a [`WlSurface`] of a window out of a set of outputs
```

Comment:
The docs should mention (and maybe the name should be changed to reflect this), that this essentially tries to find the preferred frame rate for a given surface.

For shell-interactions "primary"-output might instead be the output of most overlap, independent of the refresh rate. Using this function for those cases would result in a bad user experience preferring high-refresh outputs over anything else.

That said the function is perfectly fine (and quite clever) for this use-case, it is just not apparent from the naming and documentation, what criteria are used to define "primary".

Reply:
yeah, valid points. I will try to find a different name and document what this functiont does. Maybe it would be a good idea to also provide a function that returns the output of most overlap while at it?

Reply:
> Maybe it would be a good idea to also provide a function that returns the output of most overlap while at it?

Likely, I have implemented such functionality in both anvil and cosmic-comp at some point, so there is likely a useful helper function

Reply:
renamed to `window_surface_preferred_output` and mentioned the refresh rate in the docs.
also added `window_surface_output_most_overlap` for shell-interactions